open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/utils/App2AppAuthUtils.java

```diff
@@ -0,0 +1,180 @@
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.exception.PushDeviceHandlerClientException;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.exception.PushDeviceHandlerServerException;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.model.Device;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.core.common.AbstractUserStoreManager;
+import org.wso2.carbon.user.core.service.RealmService;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Utils class for Authentication related logic implementations.
+ */
+public class App2AppAuthUtils {
+
+    /**
+     * Retrieves an authenticated user object based on the provided subject identifier.
+     *
+     * @param subjectIdentifier the subject identifier used to retrieve the authenticated user
+     * @return an AuthenticatedUser object representing the authenticated user
+     */
+    public static AuthenticatedUser getAuthenticatedUserFromSubjectIdentifier(String subjectIdentifier) {
+
+            return AuthenticatedUser.createLocalAuthenticatedUserFromSubjectIdentifier(subjectIdentifier);
+    }
+
+    /**
+     * Retrieves the user realm associated with the provided authenticated user.
+     *
+     * @param authenticatedUser the authenticated user for whom to retrieve the user realm
+     * @return the user realm associated with the authenticated user, or null if the user is not authenticated
+     * @throws UserStoreException if an error occurs while retrieving the user realm
+     */
+    public static UserRealm getUserRealm(AuthenticatedUser authenticatedUser) throws UserStoreException {
+
+        UserRealm userRealm = null;
+
+        if (authenticatedUser != null) {
+            String tenantDomain = authenticatedUser.getTenantDomain();
+            int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+            RealmService realmService = IdentityExtensionsDataHolder.getInstance().getRealmService();
+            userRealm = realmService.getTenantUserRealm(tenantId);
+        }
+
+        return userRealm;
+    }
+
+    /**
+     * Retrieves the user ID associated with the provided username from the specified user realm.
+     *
+     * @param username the username for which to retrieve the user ID
+     * @param userRealm the user realm from which to retrieve the user ID
+     * @return the user ID associated with the username
+     * @throws UserStoreException if an error occurs while retrieving the user ID
+     */
+    public static String getUserIdFromUsername(String username, UserRealm userRealm) throws UserStoreException,
+            OpenBankingException  {
+
+        if (userRealm != null) {
+            AbstractUserStoreManager userStoreManager = (AbstractUserStoreManager) userRealm.getUserStoreManager();
```

Comment:
Why we're casting this one?

Reply:
- UserStoreManager is a interface
- userRealm returns UserStoreManager
- AbstractUserStoreManager implements UserStoreManager
- method used there (getUserIDFromUserName()) is not implemented in UserStoreManager
- Hence downcasting is required.